src/physics/sitetype.jl

```diff
@@ -204,35 +204,48 @@ op(s::Vector{<:Index},
 #
 #---------------------------------------
 
-state(s::Index,n::Integer) = s[n]
+@eval struct StateName{Name}
+  (f::Type{<:StateName})() = $(Expr(:new, :f))
+end
+
+StateName(s::AbstractString) = StateName{SmallString(s)}()
+StateName(s::SmallString) = StateName{s}()
+name(::StateName{N}) where {N} = N
+
+macro StateName_str(s)
+  StateName{SmallString(s)}
+end
+
+state(::SiteType,::StateName) = nothing
+state(::SiteType,::AbstractString) = nothing
 
 function state(s::Index,
-               str::String)::IndexVal
-  use_tag = 0
-  nfound = 0
-  for n=1:length(tags(s))
-    SType = SiteType{tags(s)[n]}
-    if hasmethod(state,Tuple{SType,AbstractString})
-      use_tag = n
-      nfound += 1
-    end
+               name::AbstractString)::IndexVal
+  Ntags = max(1,length(tags(s))) # use max here in case of no tags
+                                 # because there may still be a
+                                 # generic case such as name=="Id"
+  stypes  = [SiteType(tags(s)[n]) for n in 1:Ntags]
+  sname = StateName(SmallString(name))
+
+  # Try calling state(::SiteType"Tag",::StateName"Name")
+  for st in stypes
+    res = state(st,sname)
+    !isnothing(res) && return s(res)
   end
-  if nfound == 0
-    throw(ArgumentError("Overload of \"state\" function not found for Index tags $(tags(s))"))
-  elseif nfound > 1
-    throw(ArgumentError("Multiple tags from $(tags(s)) overload the function \"state\""))
+
+  # Try calling state(::SiteType"Tag","Name")
+  for st in stypes
```

Comment:
Is this version for backwards compatibility, in case someone had already defined their own `state` definitions?

Reply:
Yes, exactly. Though I doubt anyone has. Of course it will only run in either that case or when an error is about to occur anyway, so no loss of speed except when using the older function signature.